src/helix/impl/props.cljc

```diff
@@ -72,6 +72,47 @@
 (defn merge-obj [o1 o2]
   #?(:cljs (js/Object.assign o1 o2)))
 
+(defn seq-to-class [class]
+  (if (sequential? class)
+    (->> class
+         (map str)
+         (string/join " "))
+    class))
+
+(defn clean-class [class]
+  (if (string? class)
+    (-> class
+        (string/replace #"[ \n]+" " ")
+        (string/trim))
+    class))
+
+#?(:clj
+   (defn unquote-class
+     "Handle the case of (quote '[foo bar])"
+     [class]
+     (if (and (list? class)
+              (= (first class) 'quote))
+       (-> class
+           second
+           seq-to-class
+           str)
+       class)))
+
+#?(:clj
+   (defn normalize-class [class]
+     (-> class
+         unquote-class
+         clean-class)))
```

Comment:
I think that we just want to emit a runtime check when at macro time. So the `:clj` branch here would just be:

```suggestion
   (defn normalize-class [class]
     `(normalize-class ~class))
```

If we want to do some compile-time optimizations of this, we can detect if the user passed in a literal string and omit the check:

```suggestion
   (defn normalize-class [class]
     (if (string? class)
       class
       `(normalize-class ~class))
```

If we wanted to get REAL fancy, we could try and use type inference to infer whether the value they passed can be detected as a string, but that would probably require a significant refactor of `-native-props`

Reply:
> I think that we just want to emit a runtime check when at macro time. So the :clj branch here would just be ...

The idea is the conversion of `{:class '[foo bar])` to `{:class "foo bar"}` could be done at macro time, so in the clj branch we check for that.

Reply:
I'm not sure why a user would be passing in quoted symbols anyway. But right now the runtime `normalize-class` never gets executed if they write props literally AFAICT...

Try adding a test case for `{:class (vector foo bar)}`, it needs to emit `{:class (normalize-class (vector foo bar))}` if we want it to do runtime conversion when it's not a string or quoted vector.

Reply:
To explain a little further, `-native-props` is only called at runtime if you use dynamic props, e.g. `($ "div" {& props})`. Otherwise it will only ever take the Clojure `-native-props` path, so we need to ensure that path emits calls to any runtime processing we want to do on static props.

Reply:
A good example is https://github.com/Lokeh/helix/blob/9c3a746cda5048c40fb8177de0c38158afc06116/src/helix/impl/props.cljc#L67

when the user writes `:style [{,,,} {,,,}]` it's going to follow the `:clj` branch so we need to, at macro time, emit some code to convert the seq into a JS array.

when the user passes in a dynamic map with a `:style` key that is a seq, it follows the `:cljs` branch so we just do the runtime processing of the seq into an array there.

Reply:
> I'm not sure why a user would be passing in quoted symbols anyway

`'[foo bar baz bam]` is an easier way to write `['foo 'bar 'baz 'bam]`

> Try adding a test case for {:class (vector foo bar)}, it needs to emit {:class (normalize-class (vector foo bar))} if we want it to do runtime conversion when it's not a string or quoted vector.

done.

> To explain a little further, -native-props is only called at runtime if you use dynamic props ...

I can see that more clearly now. Thanks!